### PR TITLE
input managers

### PR DESCRIPTION
**Pull request by jamiedemaria**
### Summary & Motivation
First pass at using input managers on ops based on our meeting yesterday

`graph_job_op_toys/input_managers.py` is the most important file for getting a feel for what using this might look like to a user. The rest of the files are just making it so `In` can have a specific input manager (and it's a bit hacky right now). I still need to do a bit of work so that you dont have to specify the input manager as a required resource on the op using it, but you can imagine that that's taken care of. 

I went with the version where a user would write a new IO manager that extends the one being used by upstream ops and just replace the `load_input` function. I can also write up an example of what it might look like to have an input manager that doesnt extend from another IO manager if that would be helpful. 

I'm interested to see what you think of this approach, how it looks to the user, if it makes sense etc. 

Update: once the general API approach looks good we'll need to:
- [x] unit test
- [x] potentially merge root_input_manager into input_manager and add deprecation warnings 
- [x] docs

Update: starting trying to merge root input manager and input manager. So far it requires the following assumptions
1. any input manager can also be a root input manager and vis versa - there are a couple cases where we need to differentiate what happens to an input if it has a root input manager or an input manager and for the case of this exercise it was easier to assume that anything set for the input_manager_key could do either 

I updated the `input_managers.py` example file to load the input to one of the ops as a root input - you can see that in the NumpyCSVIOManager there is a case to determine if there is an upstream output and provide a root input value if there is now upstream output

### CODE REVIEW

File: python_modules/dagster-test/dagster_test/graph_job_op_toys/input_managers.py [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+class PysparkCsvInputManager(PandasCsvIOManager):
+    def load_input(self, context: InputContext) -> pyspark.DataFrame:
+        file_path = self._get_path(context.upstream_output)
+        spark = pyspark.SparkSession.builder.getOrCreate()
+        df = spark.read.csv(file_path)
+        return df
+
+
+@io_manager(config_schema={"base_dir": Field(Noneable(str), default_value=None, is_required=False)})
+def pyspark_input_manager(init_context):
```
Comment (jamiedemaria) [inline python_modules/dagster-test/dagster_test/graph_job_op_toys/input_managers.py:53]: had issues passing pyspark dataframe around (never used pyspark before so i don't really know what im doing with it) so switched to using numpy for the sake of getting the example working. keeping the pyspark io manager around just in case i want to use it again, but you can ignore it for now

**Review by sryza — COMMENTED** [review]
The implementation looks right to me

Not necessarily the end of the world, but "InputManager" in the name PySparkCsvInputManager feels a little weird to me, given that it's actually a full IO manager.

A case to consider is when the upstream asset doesn't have any IO manager at all, i.e. when it has a `Nothing` output. E.g.
```
@op(out=Out(Nothing, name="users"))
def my_op():
    create_db_connection().execute("create table users as ...")

@op(ins={"users": In(input_manager_key=...)}
def downstream_op(users: DataFrame):
    ...
```

In that case, there wouldn't be an upstream IO manager to extend. The most straightforward way to define an input manager might look something like the way you define an root_input_manager (no need to make a class, just decorate a function).

**Comment by jamiedemaria** [comment]
@sryza fair point on InputManager being a misleading name. I've got no issue having the recommended name be IOManager instead. I'm setting up some time w Owen next week to get up to speed on the dbt (or other Nothing output) use case so I can mock that up too

File: python_modules/dagster/dagster/core/definitions/input.py [hunk 2]
```diff
@@ -405,6 +415,7 @@ def __new__(
             metadata=check.opt_dict_param(metadata, "metadata", key_type=str),
             asset_key=check.opt_inst_param(asset_key, "asset_key", (AssetKey, FunctionType)),
             asset_partitions=asset_partitions,
+            input_manager_key=input_manager_key,
```
Comment (jamiedemaria) [inline python_modules/dagster/dagster/core/definitions/input.py:418]: add check opt string

File: python_modules/dagster/dagster/core/definitions/dependency.py [hunk 3]
```diff
@@ -258,10 +258,12 @@ def get_resource_requirements(
...
-                if isinstance(
-                    requirement, InputManagerRequirement
-                ) and outer_container.dependency_structure.has_deps(
-                    SolidInputHandle(self, solid_def.input_def_named(requirement.input_name))
+                if (
+                    isinstance(requirement, InputManagerRequirement)
+                    and outer_container.dependency_structure.has_deps(
+                        SolidInputHandle(self, solid_def.input_def_named(requirement.input_name))
+                    )
+                    and requirement.root_input
```
Comment (jamiedemaria) [inline python_modules/dagster/dagster/core/definitions/dependency.py:266]: i need a way to differentiate between inputs with a root input manager vs ones with a specific input manager. right now doing it by adding an attribute to InputManagerRequirement, but open to other ways of doing this

File: python_modules/dagster/dagster/core/definitions/run_config.py [hunk 4]
```diff
@@ -194,7 +194,9 @@ def get_inputs_field(
     for name, inp in solid.definition.input_dict.items():
         inp_handle = SolidInputHandle(solid, inp)
         has_upstream = input_has_upstream(dependency_structure, inp_handle, solid, name)
-        if asset_layer.asset_key_for_input(handle, name) and not has_upstream:
+        if inp.input_manager_key and has_upstream:
```
Comment (jamiedemaria) [inline python_modules/dagster/dagster/core/definitions/run_config.py:197]: double check that this condition makes sense and makes sense wrt the rest of the options

File: python_modules/dagster/dagster/core/definitions/solid_definition.py [hunk 5]
```diff
@@ -323,6 +331,7 @@ def get_resource_requirements(
                         key=io_manager_key,
                         node_description=node_description,
                         input_name=input_def.name,
+                        root_input=False,  # TODO maybe need to be true
```
Comment (jamiedemaria) [inline python_modules/dagster/dagster/core/definitions/solid_definition.py:334]: if we keep the `root_input` attribute in InputManagerRequirement, need to make sure that this is set correctly for the asset case

File: python_modules/dagster/dagster/core/definitions/dependency.py (hunk 3 above)
Comment (dpeng817) [reply python_modules/dagster/dagster/core/definitions/dependency.py:266]: This is fine IMO. Other option is another subclass but that seems excessive.

File: python_modules/dagster/dagster/core/execution/plan/inputs.py [hunk 6]
```diff
@@ -440,18 +440,30 @@ def load_input_object(
...
+            )
+            input_manager = step_context.get_io_manager(source_handle)
+            check.invariant(
+                isinstance(input_manager, IOManager),
+                f'Input "{input_def.name}" for step "{step_context.step.key}" is depending on '
+                f'the manager of upstream output "{source_handle.output_name}" from step '
+                f'"{source_handle.step_key}" to load it, but that manager is not an IOManager. '
+                f"Please ensure that the resource returned for resource key "
+                f'"{manager_key}" is an IOManager.',
+            )
```
Comment (dpeng817) [inline python_modules/dagster/dagster/core/execution/plan/inputs.py:466]: why does the input manager need to be an IOManager here? Shouldn't it really just need to implement `IInputManager`?

Comment (jamiedemaria) [reply python_modules/dagster/dagster/core/execution/plan/inputs.py:466]: good catch - i think just a copy paste that i missed

File: python_modules/dagster/dagster/core/definitions/solid_definition.py (hunk 5 above)
Comment (jamiedemaria) [reply python_modules/dagster/dagster/core/definitions/solid_definition.py:334]: @dpeng817 I want to make sure setting `root_input=False` makes sense wrt the condition here https://github.com/dagster-io/dagster/pull/8170/files#r891736416

**Review by dpeng817 — CHANGES_REQUESTED** [review]
A bit out of the loop on this one, but a few things;
- Would be great to see some tests, cases that come to mind; where both root_manager and input_manager are provided, where upstream io manager key differs.
- Why exactly does input_manager_key need to be an IOManager? Is it just because we don't have a convenient way of specifying an InputManager only? 
- In terms of functionality, what's the difference between input manager and root manager? It seems to me that specifying an input manager is functionally equivalent to root manager, except that you don't have to specify `handle_output` for root managers. Which makes me wonder if a more holistic change could work here that merges functionality between `root_manager_key` and `input_manager_key` - something along the lines of `root_input_manager` just creating an IInputManager instead of needing a whole different definition type, and then any IInputManager could be passed to `input_manager_key`.

**Comment by jamiedemaria** [comment]
> * Would be great to see some tests, cases that come to mind; where both root_manager and input_manager are provided, where upstream io manager key differs.

Yep, tests to come, wanted to get the general API tamped down before spending a bunch of time on tests i might need to rewrite

> * Why exactly does input_manager_key need to be an IOManager? Is it just because we don't have a convenient way of specifying an InputManager only?

One of the things we discussed for input managers is this issue where the method for getting the path in an input manager  could become out of sync with the method for getting the path for the IO manager of the output it's loading. One way around this would to be to recommend that input managers are subclasses of the output manager they are associated with and they just update the `load_input` function. An option would be to change any type checking to be IInputManager instead of IOManager and just document the recommended way to use input managers. 

> * In terms of functionality, what's the difference between input manager and root manager? It seems to me that specifying an input manager is functionally equivalent to root manager, except that you don't have to specify `handle_output` for root managers. Which makes me wonder if a more holistic change could work here that merges functionality between `root_manager_key` and `input_manager_key` - something along the lines of `root_input_manager` just creating an IInputManager instead of needing a whole different definition type, and then any IInputManager could be passed to `input_manager_key`.

This is another thing we discussed. potentially input manager would subsume root_input manager. once the input manager API is looking good we could merge root_input_manager into input_manager and throw a deprecation warning on root_input_manager

**Comment by jamiedemaria** [comment]
spending a bit of time thinking about removing root_input_manager and the first potential issue i'm seeing is that we may not be able to support the root input manager with subselection documented [here](https://docs.dagster.io/concepts/io-management/unconnected-inputs#using-a-root-input-manager-with-subselection-) 

the case where this would be an issue is if we had fully deprecated the `root_manager_key` in `In` then if a user wanted to specify an input manager for normal loading AND a root input manager for loading if the op is being run with no upstream output there might be an issue. maybe this would be solved by writing a custom input manager that handles both cases but that could end up with a  user writing.an input manager for each op.

 i guess now that i'm writing this, that's sort of what happens with the root input managers anyway so maybe this isn't a big deal

**Comment by jamiedemaria** [comment]
updated to look at merging root_input_manager into input_manager. a bit rough right now, but gives the general idea. see the addition to the main pr description for more info

**Comment by jamiedemaria** [comment]
thinking more about merging/deprecating root input managers - after going through the exercise of merging root input managers into input managers i think we run into an issue with the deprecation of the `@root_input_manager` decorator. 

Originally i was thinking that we would merge root input managers and normal input managers on our end and just put a deprecation warning on the use of the `root_manager_key` arg, but functions decorated by `root_input_manager` stop working if we merge the two types of input managers on the backend. so i think it would be best to keep root_input_managers as they are for right now, add in the deprecation warnings, and then rip out all of the functionality when we deprecate for real.

planning to revert my previous commit and then out in deprecation warnings

File: docs/content/concepts/io-management/unconnected-inputs.mdx [hunk 7]
```diff
@@ -103,6 +101,130 @@ my_job.execute_in_process(
...
+
+
+@job(resource_defs={"load_input_manager": table_1_manager})
+def load_table_job():
+    my_op()
+```
+
+Setting the `input_manager_key` on an `In` controls how that input is loaded.
+
+### Providing per-input config to an input manager
```
Comment (jamiedemaria) [inline docs/content/concepts/io-management/unconnected-inputs.mdx:137]: need to actually support this - working on it now

File: docs/content/concepts/io-management/io-managers.mdx [hunk 8]
```diff
@@ -174,6 +174,66 @@ local_assets = with_resources(
 )
 ```
 
+### Customizing how assets are loaded downstream
```
Comment (jamiedemaria) [inline docs/content/concepts/io-management/io-managers.mdx:177]: i feel pretty uncertain about where this is placed in the page. Thematically it fits with the asset IO section, but it relies on the user know how IO Managers are defined, which isn't covered until the next section. I *could* shorten this to be more like the per-output section, and then show the full code snippet in an example, but i'd be worried that some important context (that input managers need to have some knowledge of the how the output was stored) would be missed by users who don't look at the example. what do you think?

File: docs/content/concepts/io-management/io-managers.mdx [hunk 9]
```diff
@@ -391,6 +451,88 @@ def my_io_manager(_):
     return MyIOManager()
 ```
 
+### Per-input IO Manager
```
Comment (jamiedemaria) [inline docs/content/concepts/io-management/io-managers.mdx:454]: similarly unsure where this section should be (same concerns as the per input io management for assets)

File: docs/content/concepts/io-management/unconnected-inputs.mdx [hunk 10]
```diff
@@ -9,16 +9,14 @@ description:
...
 
 ## Overview
 
 Ops in a job may have input definitions that don't correspond to the outputs of upstream ops. You can provide values for these inputs in a few different ways. Dagster checks each, in order, and uses the first that's available:
 
-- **Root Input Manager (experimental)** - If the input to a job comes from an external source, such as a table in a database, often it makes sense to define a resource that's responsible for loading it. This makes it easy to swap out implementations in different jobs and mock it in tests. A <PyObject module="dagster" object="RootInputManager" />, which can be referenced from <PyObject module="dagster" object="In" />s, is a resource that loads unconnected inputs.
+- **Input Manager (experimental)** - If the input to a job comes from an external source, such as a table in a database, often it makes sense to define a resource that's responsible for loading it. This makes it easy to swap out implementations in different jobs and mock it in tests. A special IOManager, which can be referenced from <PyObject module="dagster" object="In" />s, can be used to loads unconnected inputs.
```
Comment (jamiedemaria) [inline docs/content/concepts/io-management/unconnected-inputs.mdx:19]: not sure if the "input manager" concept will be experimental because it's just an IO manager which is stable

File: docs/content/concepts/io-management/unconnected-inputs.mdx [hunk 11]
```diff
@@ -103,6 +101,143 @@ my_job.execute_in_process(
...
+
+Now, when running the full job, `op2`'s input will be loaded using the IO manager on the output of `op1`. When running the job subset, `op2`'s input has no upstream output, so `table_1` will be loaded.
+
+```python file=/concepts/io_management/input_managers.py startafter=start_execute_subselection endbefore=end_execute_subselection
+my_subselection_job.execute_in_process(
+    op_selection=["op2"],
+)
+```
+
+## Root Input Managers
```
Comment (jamiedemaria) [inline docs/content/concepts/io-management/unconnected-inputs.mdx:237]: do we want to keep the docs for root input managers around as reference for people who already wrote them? or do we expect them to just pull up an older version of the docs?

File: docs/content/concepts/io-management/io-managers.mdx [hunk 12]
```diff
@@ -174,6 +174,66 @@ local_assets = with_resources(
 )
 ```
 
+### Customizing how assets are loaded downstream
+
+In some cases, you may find that you want to write some custom logic for how an upstream asset is loaded in a downstream asset. For example, if an asset is set to store and load as a pandas DataFrame, but you want to load it with PySpark in a downstream asset, it may make more sense to write a new loader for the downstream asset than to update the upstream asset's IO Manager to handle loading as a pandas DataFrame or a PySpark dataframe.
```
Comment (sryza) [inline docs/content/concepts/io-management/io-managers.mdx:179]: There's some inconsistency throughout these docs in the capitalization for Pandas and DataFrame. I think we should capitalize both as capitalized in the previous sentence.

File: docs/content/concepts/io-management/io-managers.mdx (hunk 8 above)
Comment (sryza) [reply docs/content/concepts/io-management/io-managers.mdx:177]: Would it possibly make sense to introduce setting input managers before introducing how to define them? Similar to what we do with IO managers?

File: docs/content/concepts/io-management/io-managers.mdx [hunk 13]
```diff
@@ -391,6 +451,88 @@ def my_io_manager(_):
     return MyIOManager()
 ```
 
+### Per-input IO Manager
+
+In some cases you may find that you need to load an input in a way other than the `load_input` function of the corresponding output's IO Manager. For example, let's say Team A has an op that returns an output as a pandas Dataframe and specifies an IO manager that knows how to store and load pandas DataFrames. Your team is interested in using this output for a new op, but you are required to use PySpark to analyze the data. Unfortunately, you don't have permission to modify Team A's IO Manager to support this case. Instead, you can specify an input manager on your op that will override some of the behavior of Team A's IO Manager.
```
Comment (sryza) [inline docs/content/concepts/io-management/io-managers.mdx:456]: In the rest of this doc, it looks like we typically write "IO manager" instead of "IO Manager".

**Review by sryza — COMMENTED** [review]
This is looking great - my main outstanding question: do we want to enable some way of directly defining an input manager, similar to how you can directly define a root_input_manager?

File: docs/content/concepts/io-management/io-managers.mdx (hunk 8 above)
Comment (jamiedemaria) [reply docs/content/concepts/io-management/io-managers.mdx]: rearranged the docs section, lmk how it reads now

**Comment by sryza** [comment]
> it could be weird to provide a simple @input_manager decorator and not provide an @output_manager

IMO this is not that weird. If we thought people would use an @output_manager decorator, I'd say we should add one, but I haven't seen much demand.

> I think there's also the question of if @input_manager would decorate a function that returns the input (like @root_input_manager) or returns a class that follows the InputManager interface (similar to @io_manager).

My bent would be towards decorating a function that returns the input, because it's more ergonomic / less boilerplate. Unless there are compelling reasons for returning an InputManager instance?

**Comment by jamiedemaria** [comment]
Ok let me see what it would be like having an `@input_manager` that would decorate a plain function. 

when i did a first pass at deprecating `root_input_manager` i tried to merge the functionality into `input_manager` so that we'd just have to remove arg and decorator when it came time to deprecate. iirc i ran into issues trying to handle both the decorated function and an InputManager class when figuring out how to load an input. I don't remember the specifics, but may run into the same issue adding `@input_manager`. we'll see!

File: docs/content/concepts/io-management/io-managers.mdx [hunk 14]
```diff
@@ -232,6 +269,28 @@ def my_job():
     op_2(op_1())
 ```
 
+### Per-input IO manager
+
+Just as with the inputs to assets, the inputs to ops can be loaded using custom logic. To set an IO manager for a particular input, use the `input_manager_key` argument on <PyObject module="dagster" object="In" />.
```
Comment (sryza) [inline docs/content/concepts/io-management/io-managers.mdx:274]: I think it's worth clarifying that this is only for when you want to override the IO manager from the upstream output.

File: docs/content/concepts/io-management/io-managers.mdx [hunk 15]
```diff
@@ -278,6 +337,88 @@ class MyPartitionedIOManager(IOManager):
...
+
+Since the method for loading an input is directly affected by the way the corresponding output was stored, we recommend defining your input managers as subclasses of existing IO managers and just updating the `load_input` method. In this example, we load an input as a NumPy array rather than a Pandas DataFrame by writing the following:
+
+```python file=/concepts/io_management/input_managers.py startafter=start_plain_input_manager endbefore=end_plain_input_manager
+# in this case PandasIOManager is an existing IO Manager
+class MyNumpyLoader(PandasIOManager):
```
Comment (sryza) [inline docs/content/concepts/io-management/io-managers.mdx:348]: Would the `@input_manager` decorator make sense in this case?

**Review by sryza — COMMENTED** [review]
Looking great! Had two small remaining comments.

File: docs/content/concepts/io-management/io-managers.mdx (hunk 15 above)
Comment (jamiedemaria) [reply docs/content/concepts/io-management/io-managers.mdx:348]: i think in this first version it could be a plain input manager, but in the better version of the example it has to be an IO Manager because it shares the `_get_path` function defined by the PandasIOManager. I think the example is more streamlined if we just start off with an IO manager instead of going from input manager to io manager

File: examples/modern_data_stack_assets/modern_data_stack_assets/assets.py [hunk 16]
```diff
@@ -23,10 +23,13 @@
...
-@asset(compute_kind="python")
-def order_forecast_model(daily_order_summary: pd.DataFrame) -> Any:
+@asset(
+    compute_kind="python",
+    ins={"daily_order_summary": AssetIn(namespace="public", input_manager_key="numpy_io_manager")},
+)
+def order_forecast_model(daily_order_summary: np.ndarray) -> Any:
     """Model parameters that best fit the observed data"""
-    df = daily_order_summary
+    df = pd.DataFrame(daily_order_summary, columns=["order_date", "total_value", "num_orders"])
```
Comment (OwenKephart) [inline examples/modern_data_stack_assets/modern_data_stack_assets/assets.py:33]: tiny nitpick here, but right now we take in daily_order_summary as a numpy thing, then convert it to pandas, then convert it right back to a numpy thing again on line 35, which is pretty roundabout.

this is actually a perfect usecase for input managers (we want the thing as numpy in our calculations, not pandas, so it's great that we can load the input as pandas in some places and numpy in others), so I think it would be best if we did not do this conversion to pandas at all.

Instead we can slightly change the optimize.curve_fit call to just work directly with numpy. I believe that would look like `f=model_funcm xdata=daily_order_summary[0, :], ydata=daily_order_summary[2, :], p0=[10, 100]` (but I might have gotten the axes flipped).

**Review by dpeng817 — APPROVED** [review]
